tests/h/views/activity_test.py

```diff
@@ -117,17 +117,17 @@ class TestGroupSearchController(object):
 
     """Tests unique to GroupSearchController."""
 
-    def test_renders_join_template_when_no_read_permission(self, controller, pyramid_request, group):
+    def test_renders_join_template_when_no_read_permission(self, controller, pyramid_request, group, factories):
         """When the request has no read permission but join, it should render the join template."""
 
         def fake_has_permission(permission, context=None):
             if permission == 'read':
                 return False
-            if permission == 'join':
+            elif permission == 'join':
```

Comment:
Does changing `if` to `elif` here make an difference?

Reply:
We probably want to refactor these tests to get rid of all the `fake_has_permission`s. What is the view actually supposed to do? For private groups it's supposed to show a login page if the user isn't logged in, a join page if the user is logged in but isn't a member, and the group's page if the user is a member. For open and restricted groups it's slightly different (there's no join page or login page, it just shows the groups page in all three instances, but all three instances should still be tested).

So I think the tests should work by, for example, setting a group, and a user who is a member of that group, and a group page request from that user, and then test what happens.

The tests shouldn't mock `has_permission()`.

Reply:
Making these improvements to the tests first would make it safer to refactor the code after